### PR TITLE
add BOM

### PR DESCRIPTION
**Pull request by elharo**
@garrettjonesgoogle Is this what you were thinking of?

### CODE REVIEW

**Comment by garrettjonesgoogle** [comment]
Yep!

**Comment by xfxyjwf** [comment]
Probably silly question, but what is this?

**Comment by xfxyjwf** [comment]
@ejona86 Can you also help take a look?

**Comment by BSBandme** [comment]
I just went through go/cloud-java-bom, some questions:
- Should we add BOM to java lite?
- Are the follow up things just changing parent POM to include this BOM to \<dependecyManagement\> and \<modules\> after this submitted?
-  Do we need to add \<build\> to BOM for build tesing? Or can parent building cover this?

**Comment by BSBandme** [comment]
@ejona86 Do you have any comments for this? This LGTM, I think it's OK to merge?

**Comment by elharo** [comment]
I was trying to make this more standalone so I didn't inherit from the parent pom, but maybe it should. Give me a minute to explore that and see if it has other implications.

I added the bom module to the parent pom.

**Comment by garrettjonesgoogle** [comment]
The BOM shouldn't inherit from the parent POM (see internal doc). It does indeed need to be listed as a module in the parent POM module though.

**Comment by elharo** [comment]
@saturnism Care to weigh in on whether we should inherit the parent pom here? I'm leaning no, but might be nice to get the reasons on record.

**Comment by garrettjonesgoogle** [comment]
For thoroughness:

1. protobuf/java/pom.xml should import the BOM in a dependency management section
2. In protobuf/java/util/pom.xml, the dependency on protobuf-java should have its version removed
3. We should make sure that protobuf's version bumping process before release correctly bumps the BOM's version

**Comment by elharo** [comment]
1. Check.
2. Can't seem to get this one to work.

File: java/pom.xml [hunk 1]
```diff
@@ -61,6 +61,12 @@
 
   <dependencyManagement>
     <dependencies>
+      <dependency>
+        <groupId>com.google.protobuf</groupId>
+        <artifactId>protobuf-bom</artifactId>
+        <version>${project.version}</version>
+        <type>pom</type>
+      </dependency>
```
Comment (garrettjonesgoogle) [inline java/pom.xml]: I believe this also needs to have `<scope>import</scope>`

Comment (elharo) [reply java/pom.xml]: That fixed it. Thanks. PTAL @garrettjonesgoogle @ejona86

**Comment by BSBandme** [comment]
Hi Eric, any updates? @ejona86

**Comment by ejona86** [comment]
This still looks broken for the release process. Uploading to sonatype (as a SNAPSHOT would be fine) needs to be tested with this. I'm pretty certain it is broken, since the bom has no distributionManagement nor nexus-staging-maven-plugin. I honestly don't know the configuration is necessary for protobuf, but what's here seem insufficient.

**Comment by garrettjonesgoogle** [comment]
We have distributionManagement and nexus-staging-maven-plugin configured in google-cloud-bom (https://github.com/googleapis/google-cloud-java/blob/master/google-cloud-bom/pom.xml), so I'm pretty sure what @ejona86 is saying is correct.

**Comment by garrettjonesgoogle** [comment]
@elharo : In an offline conversation with @ejona86 , he said to make sure that `mvn install` works at a minimum (and I would suggest that you also make sure that the BOM can be used successfully in google-cloud-java on your machine also as a test, even though we would not check that in), then he can make sure someone takes care of the snapshot test.

**Comment by elharo** [comment]
Anyone else seeing these failures? The SurefireBooterForkException problem looks like one we saw earlier this week in cloud-opensource-java and may require unrelated changes. 

```
[ERROR] Errors: 
[ERROR]   AnyTest.testAnyGeneratedApi » StackOverflow
[ERROR]   DiscardUnknownFieldsTest.testProto3 » StackOverflow
[ERROR]   FieldPresenceTest.testFieldPresence » StackOverflow
[ERROR]   GeneratedMessageTest.testDeserializeWithClassField » StackOverflow
[ERROR]   GeneratedMessageTest.testDeserializeWithoutClassField » StackOverflow
[ERROR]   MapTest.testCount » StackOverflow
[ERROR]   MapTest.testGettersAndSetters » StackOverflow
[ERROR]   MapTest.testInitFromFullyDynamicMessage » StackOverflow
[ERROR]   MapTest.testInitFromPartialDynamicMessage » StackOverflow
[ERROR]   MapTest.testMergeFrom » StackOverflow
[ERROR]   MapTest.testNestedBuilderOnChangeEventPropagation » StackOverflow
[ERROR]   MapTest.testParseError » StackOverflow
[ERROR]   MapTest.testSerializeAndParse » StackOverflow
[ERROR]   MapTest.testUnknownEnumValues » StackOverflow
[ERROR]   MapTest.testUpdateMapValues » StackOverflow
[INFO] 
[ERROR] Tests run: 632, Failures: 0, Errors: 15, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Protocol Buffers [BOM] ............................. SUCCESS [  0.339 s]
[INFO] Protocol Buffers [Parent] .......................... SUCCESS [  0.019 s]
[INFO] Protocol Buffers [Core] ............................ FAILURE [05:40 min]
[INFO] Protocol Buffers [Util] ............................ SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 05:44 min
[INFO] Finished at: 2018-11-01T10:44:33-04:00
[INFO] Final Memory: 32M/715M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.0:test (default-test) on project protobuf-java: There are test failures.
[ERROR] 
[ERROR] Please refer to /Users/elharo/protobuf/java/core/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date]-jvmRun[N].dump, [date].dumpstream and [date]-jvmRun[N].dumpstream.
[ERROR] The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
[ERROR] Command was /bin/sh -c cd /Users/elharo/protobuf/java/core && /Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/bin/java -jar /Users/elharo/protobuf/java/core/target/surefire/surefirebooter6394566698081996952.jar /Users/elharo/protobuf/java/core/target/surefire 2018-11-01T10-39-43_040-jvmRun1 surefire1038890618998964833tmp surefire_08369381138253167683tmp
[ERROR] Error occurred in starting fork, check output in log
[ERROR] Process Exit Code: 1
[ERROR] Crashed tests:
[ERROR] com.google.protobuf.IsValidUtf8Test
[ERROR] org.apache.maven.surefire.booter.SurefireBooterForkException: The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
[ERROR] Command was /bin/sh -c cd /Users/elharo/protobuf/java/core && /Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk/Contents/Home/jre/bin/java -jar /Users/elharo/protobuf/java/core/target/surefire/surefirebooter6394566698081996952.jar /Users/elharo/protobuf/java/core/target/surefire 2018-11-01T10-39-43_040-jvmRun1 surefire1038890618998964833tmp surefire_08369381138253167683tmp
[ERROR] Error occurred in starting fork, check output in log
[ERROR] Process Exit Code: 1
[ERROR] Crashed tests:
[ERROR] com.google.protobuf.IsValidUtf8Test
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:671)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:533)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:278)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:244)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeProvider(AbstractSurefireMojo.java:1194)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:1022)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute(AbstractSurefireMojo.java:868)
[ERROR] 	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:154)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:146)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
[ERROR] 	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
[ERROR] 	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:309)
[ERROR] 	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:194)
[ERROR] 	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:107)
[ERROR] 	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:993)
[ERROR] 	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:345)
[ERROR] 	at org.apache.maven.cli.MavenCli.main(MavenCli.java:191)
[ERROR] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.lang.reflect.Method.invoke(Method.java:498)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
[ERROR] 
[ERROR] -> [Help 1]
[ERROR] 

```

**Comment by ejona86** [comment]
@elharo, I don't see a failure. I'm using Maven 3.5.2 and Java 1.8.

**Comment by ejona86** [comment]
To test this properly, you need to use the "release" profile. This worked for me:

```
mvn install -Prelease -DskipTests -Dadditionalparam="-Xdoclint:none"
```

I've confirmed for actual deployment it uses nexus-staging-maven-plugin. That is possible to test, even without access to the repository, although you will still need a sonatype login (it just uses it to read some configuration information). I got this to work:

```
mvn deploy -Prelease -DskipRemoteStaging -DskipTests -Dadditionalparam="-Xdoclint:none" -DaltStagingDirectory=/tmp/blah-deploy -Dmaven.install.skip
```

**Comment by ejona86** [comment]
Note: If you do mvn install, make sure to `rm -r ~/.m2/repository/com/google/protobuf/` afterward, because the version in master isn't a SNAPSHOT, and so the 2.6.1 version is "polluted" by the local build.

**Comment by elharo** [comment]
OK. After adding a `deploymentManagement` section I got as far as:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-deploy-plugin:2.7:deploy (default-deploy) on project protobuf-bom: Failed to deploy artifacts: Could not transfer artifact com.google.protobuf:protobuf-bom:pom:3.6.1 from/to sonatype-nexus-staging (https://oss.sonatype.org/service/local/staging/deploy/maven2/): Access denied to: https://oss.sonatype.org/service/local/staging/deploy/maven2/com/google/protobuf/protobuf-bom/3.6.1/protobuf-bom-3.6.1.pom, ReasonPhrase: Forbidden. -> [Help 1]
[ERROR] 
```

Is this WAI given that I was using my personal Nexus account that presumably does not and should not have permission to deploy this?

**Comment by elharo** [comment]
Ping @garrettjonesgoogle @ejona86

**Comment by garrettjonesgoogle** [comment]
@elharo - @ejona86 mentioned that `nexus-staging-maven-plugin` is missing and I still don't see it; is that not necessary?

**Comment by elharo** [comment]
It doesn't seem to be so far as I can tell. However it may be that my lack of access kept me from hitting the point where we need it. 

Following the docs, it looks like it's an alternative to the maven-deploy-plugin being used here.  The maven-deploy-plugin is what some other BOMs are using, but I can check to see if we can translate this to the nexus-staging-maven-plugin instead.

**Comment by garrettjonesgoogle** [comment]
I think the best thing to do is just copy the plugins out of protobuf's root xml, which uses `nexus-staging-maven-plugin` - see https://github.com/protocolbuffers/protobuf/blob/master/java/pom.xml#L188 for example.

File: java/bom/pom.xml [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        <artifactId>protobuf-java-util</artifactId>
+        <version>${project.version}</version>
+      </dependency>
+    </dependencies>
+  </dependencyManagement>
+
+  <build>
+    <plugins>
+      <plugin>
+        <artifactId>maven-gpg-plugin</artifactId>
```
Comment (ejona86) [inline java/bom/pom.xml]: This needs to be within a release profile, like in the parent pom. Otherwise it will always run.

**Comment by ejona86** [comment]
nexus-staging-maven-plugin is a newer replacement for maven-deploy-plugin, so yes, we want to use it. This is looking pretty good now. I noticed one small change that is necessary, but otherwise, looks good. Thanks!

**Comment by ejona86** [comment]
@elharo, thank you!

**Comment by rafi-kamal** [comment]
Sorry this is a pretty old PR, but I noticed that Protocol Buffers [BOM] has been licensed under Apache 2.0 here while the rest of Protobuf uses BSD 3-clause. @elharo do you know why is that?

**Comment by elharo** [comment]
No particular reason. Google uses the Apache license by default, although there are exceptions. If protobuf is one, please file an issue to change this. It shouldn't be a big deal.

**Comment by rafi-kamal** [comment]
Thanks!

I was about to create a PR with the change but accidentally committed it :/ This is the commit for reference: https://github.com/protocolbuffers/protobuf/commit/7efcc0466c31e18842d612b1a70cec3f1c9bfb23

**Pull request merged**